---
title: "RNAseq data analysis pipeline"
author: "Hannah"
date: "2023-10-30"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## This analysis incorporates all statistical and graphical analyses of the spat transcriptomes generated in the multiple stressor trial run in November 2022. The aim of this study was to determine if multiple stressors in the form of temperature (ocean heatwaves) and harmful algae blooms have a detrimental impact on the survival of GSM spat.

```{r Install packages, message=FALSE, warning=FALSE, collapse=T, include=FALSE}
install.packages("ggtree")
install.packages("harrietr")
install.packages("RColorBrewer")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("phyloseq")
install.packages("enrichplot")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("SummarizedExperiment")

install.packages("gplots")
install.packages("adonis")
install.packages("extrafont")
install.packages("showtext")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

install.packages("showtext")
if (!require("BiocManager", quietly = TRUE))
    install.packages("ashr")

BiocManager::install("topGO")
install.packages("gprofiler2")

BiocManager::install("clusterProfiler", version = "3.19")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
BiocManager::install(organism, character.only = TRUE)
```

### Loading libraries

```{r Loading libraries, message=FALSE, warning=FALSE, collapse=T, include=FALSE}
library(SummarizedExperiment)
#library(extrafont)
library(tidyverse); packageVersion("tidyverse")
#library("lattice"); packageVersion("lattice")
library(readr); packageVersion("readr")
#library(gplots); packageVersion("gplots")
#library(RColorBrewer); packageVersion("RColorBrewer")
library(phyloseq); packageVersion("phyloseq")
library(data.table); packageVersion("data.table")
library(vegan); packageVersion("vegan")
#library(mixOmics); packageVersion("mixOmics")
#library(compositions); packageVersion("compositions")
library(ggthemes); packageVersion("ggthemes")
#library(hrbrthemes); packageVersion("hrbrthemes")
library(viridis); packageVersion("viridis")
library(janitor); packageVersion("janitor")
library(cowplot); packageVersion("cowplot")
#library(grid); packageVersion("grid")
#library(igraph); packageVersion("igraph")
library(ggpubr); packageVersion("ggpubr")
#library(metagMisc); packageVersion("metagMisc")
#library(genefu); packageVersion("genefu")
library(pracma); packageVersion("pracma")
library(ComplexHeatmap); packageVersion("ComplexHeatmap")
library("RColorBrewer")
#library("airway")
#library("tximeta")
library(DESeq2); packageVersion("DESeq2")
library(vsn)
library(pheatmap)
library(eulerr); packageVersion("eulerr")
library(scales)
library(genefilter)
library(ggplot2)
require(GO.db)
library("eulerr")
library("ComplexHeatmap")
library("ggforce")
library("ggfortify")
library("genefilter")
library("scatterplot3d")
library("ggcorrplot")
library("corrr")
library("reshape2")
library("rafalib")
library("stringr")
library("forcats")
library("vsn")
library("genefilter")
library("ggrepel")
library("UpSetR")
library("ggtern")
library("RColorBrewer")
library("ggVennDiagram")
library("GOplot")
library("gplots")
library("lme4")
library(extrafont)
library(showtext)
library("topGO")
library("gprofiler2")
library("ashr")
library(clusterProfiler)
library(enrichplot)
library("DEGreport")
library("FactoMineR")

```

### Loading the data

```{r}
#Set wd
setwd("C:/Users/hannahg/Cawthron/Hannah G PhD Team - General/R/Spat exposure/RNAseq/RNAseq/Data")

#GO contains GO terms for some genes
GO = read.csv('musselblast2go_annot.csv')

#Count data
fcData = read.table('feature_counts_rename.txt', sep='\t', header=TRUE)
fcData

# Coldata contains sample information/metadata- must match sample IDs in the count data!!
coldata = read.csv("coldata_updated.csv")
coldata

####Get data dimensions
dim(fcData)
names(fcData)

####Rename sample headers (if not already the same)
names(fcData)[7:74] = c("HT1_0", "HT2_0", "HT3_0", "HT4_0", "HT2_4", "HT2_8", "HT3_4", "HT3_8", "HT4_4", "HT4_8", "HT5_8", "HT5_4", "HT6_4", "HT6_8",
                        "HTHH1_8", "HTHH2_4", "HTHH2_8", "HTHH3_4", "HTHH3_8", "HTHH4_8", "HTHH4_4", "HTHH5_4", "HTHH5_8", "HTHH6_4", "HTLH1_8", "HTLH2_4", "HTLH2_8",
                        "HTLH3_4", "HTLH3_8", "HTLH4_4", "HTLH4_8", "HTLH5_4", "HTLH5_8", "HTLH6_4", "LT1_0", "LT2_0", "LT3_0", "LT4_0",
                        "LT1_4", "LT2_4", "LT2_8", "LT3_4", "LT3_8", "LT4_8", "LT4_4", "LT5_4", "LT5_8", "LT6_8", "LTHH1_8", "LTHH2_4", "LTHH2_8", "LTHH3_4",
                        "LTHH3_8", "LTHH4_4", "LTHH4_8", "LTHH5_4", "LTHH5_8", "LTHH6_4", "LTLH1_8", "LTLH2_8", "LTLH2_4", "LTLH3_4", "LTLH3_8", "LTLH4_4",
                        "LTLH4_8", "LTLH5_4", "LTLH5_8", "LTLH6_4")
fcData %>% head()

####Remove annotations so table shows only gene id and sample counts
rownames(coldata) <- coldata$idrep_time
counts <- fcData[,7:74]
rownames(counts) <- fcData$Geneid
colnames(counts) <- coldata$idrep_time
head(counts)
coldata
####Check the samples in the count data table and metadata table match up
all.equal(colnames(counts),rownames(coldata))

####read in data file
head(counts)
str(counts)

head(coldata)
str(coldata)

all.equal(colnames(counts),rownames(coldata))

```


##Normalising count data

```{r Data quality assessment and quality control, include=FALSE}
# NOTE: some data normalisation has already been done on a linux preprosessing system.

####remove genes with low counts i.e. must have >5 reads in at least 4 samples
keep_genes <- rowSums( counts > 5 ) >= 4
counts_filtered <- counts[keep_genes,]
write.csv(as.data.frame(counts_filtered),file='counts_filtered.csv')

####Remove HT-5_8, HTHH-4_8, LTC1_0 and LT-4_8 due to very low gene counts from counts and metadata files
counts_filtered <- subset(counts_filtered, select = -c(HT5_8, HTHH4_8, LT1_0, LT4_8, HTHH2_8))
counts_filtered

coldata_filtered = coldata[!(row.names(coldata) %in% c("HT5_8","HTHH4_8", "LT1_0", "LT4_8", "HTHH2_8")),]
coldata_filtered

####Write in an updated metadate file
## write.csv(as.data.frame(coldata_filtered),file='coldata_filtered.csv')


####Check we haven't lost any samples in the filtering process
all.equal(colnames(counts_filtered),rownames(coldata_filtered)) #TRUE means we still have the same number of samples

#added in extra col of data for treatment.rep
coldata_filtered$treatment.rep = paste(coldata_filtered$treatment,coldata_filtered$replicate, sep=".")

#convert to factors before creating dds object
coldata_filtered$treatment=as.factor(coldata_filtered$treatment)
coldata_filtered$level=as.factor(coldata_filtered$level)
coldata_filtered$time=as.factor(coldata_filtered$time)
coldata_filtered$treatment.level.time=as.factor(coldata_filtered$treatment.level.time)
coldata_filtered$temperature=as.factor(coldata_filtered$temperature)
coldata_filtered$id=as.factor(coldata_filtered$id)
coldata_filtered$id.time=as.factor(coldata_filtered$id.time)
coldata_filtered$replicate=as.factor(coldata_filtered$replicate)
coldata_filtered$level.time=as.factor(coldata_filtered$level.time)
```

##Create dds objects so we can run multiple pairwise comparisons based
on different variables

```{r}
# Generate different dds designs based on the comparisons you want to run i.e. treatment and time
# Think about the comparisons you are interested in doing later and write the design accordingly. 
# this part of the code generates the model. You can add interaction effects by adding a semicolon i.e. design=~treatment + time:treatment:time
dds_TT <- DESeq2:::DESeqDataSetFromMatrix(countData=counts_filtered, 
                              colData=coldata_filtered, 
                              design=~treatment + time)
dds_TT
# This part of the code runs the background stats to generate a dds object. 
dds_TT <- DESeq2:::DESeq(dds_TT)

## Or time and HAB level
dds_LT <- DESeq2:::DESeqDataSetFromMatrix(countData=counts_filtered, 
                                          colData=coldata_filtered, 
                                          design=~level + time)
dds_LT

dds_LT <- DESeq2:::DESeq(dds_LT)

# HABs.time
dds_habs.time <-DESeq2:::DESeqDataSetFromMatrix(countData=counts_filtered, 
                                           colData=coldata_filtered, 
                                           design=~habs.time)
dds_habs.time

dds_habs.time<-DESeq2:::DESeq(dds_habs.time)






# Note: In order to benefit from the default settings of the package, you should put the variable of interest at the end of the formula and make sure the control level is the first level.

# # By default, R will choose a reference level for factors based on alphabetical order. Then, if you never tell the DESeq2 functions which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. There are two solutions: you can either explicitly tell results which comparison to make using the contrast argument (this will be shown later), or you can explicitly set the factors levels. In order to see the change of reference levels reflected in the results names, you need to either run DESeq or nbinomWaldTest/nbinomLRT after the re-leveling operation

## This will print the names of the comparisons you can do later. 
resultsNames(dds_TT)
resultsNames(dds_LT)
resultsNames(dds_habs.time)

# If you want, you can Subset a dds object i.e. by timepoint
dds_T4 <- subset(dds_TT, select=colData(dds_TT)$time=="T4")

## To access the normalized counts from the dds object
gene_counts_normalized <- counts(dds_HTT, normalized = TRUE) 
```
##VSD matrices
```{r}
## Data tranformation for QC
# vsd is a much faster transformation step than rlog 
# Do this to do some PCA and plot visualisations of normalised data before you move to differential gene expression testing
# Want blind=true for the vst function doesn't take the sample groups into account i.e. does transformation in unbiased manner
# Blind= true is important for doing QC steps. However, it is not appropriate when you expect that majority of genes will have large diffs in counts due to experimental design
# Use blind=false for downstream analysis
vsd <- DESeq2:::vst(dds_TT, blind=TRUE)
vsd_mat <- assay(vsd)


# r log transform- takes longer if you have a large data set
rld_TT <- rlog(dds_TT, blind = FALSE)
rld_mat_<- assay(rld_TT)

# plots the standard deviation of the transformed data, across samples, against the mean, using the shifted logarithm transformation, and the variance stabilizing transformation. The shifted logarithm has elevated standard deviation in the higher count range, while for the variance stabilized data the standard deviation is roughly constant along the whole dynamic range.

vsd_habs.time <- DESeq2:::vst(dds_habs.time, blind=FALSE)
vsd_mat_habs.time <- assay(vsd_habs.time)
# head(assay(rld_TT), 3)
# meanSdPlot(assay(rld_TT), ranks=FALSE)

# meanSdPlot(assay(vsd))
# meanSdPlot(assay(vsd_genes))


```

##Quality control/data visualisations

```{r Data distribution, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# Check distributions of samples using boxplots
boxplot(assay(vsd_TT), xlab="", ylab="Log2 counts per million",las=2,main="Normalised Distributions")
# Let's add a blue horizontal line that corresponds to the median logCPM
abline(h=median(assay(vsd_TT)), col="blue")

hist(vsd_mat)
```

###PCA plot 
#### Temp and HAB conc

```{r PCA, echo=FALSE, message=FALSE, warning=FALSE}
# Temp and HAB conc (all timepoints)
#matrix for plots
 p <- pca(vsd_mat_habs.time, metadata = coldata_filtered, removeVar = 0.1)
  
# screeplot
screeplot(p, axisLabSize = 18, titleLabSize = 22)

#biplot
 biplot(p)
 biplot(p, showLoadings = TRUE,
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)
 
#pairsplot 
pairsplot(p)

   
horn <- parallelPCA(vsd_mat_habs.time)
  horn$n
  which(cumsum(p$variance) > 80)[1]
  
# annotated PCA plot
 tiff("PCA plot_all timepoints.tiff", units="cm", width=30, height=20, res=300) 
  biplot(p,
         x= "PC1", y="PC2",
         labSize= 4,
         pointSize = 4,
    lab = paste0(p$metadata$time),
    colby = 'level',
    colkey = c("none"="#41B3A3", 'low'= "#DDCC83", 'high'= "#E27D60" ),
    hline = 0, vline = 0,
    colLegendTitle = substitute(paste(italic('A. pacificum '), "concentration")),
    legendPosition = 'none',
     axisLabSize = 12)
   dev.off()
   
   # Subset vsd matrix to T4 only
vsd_habs.time.T4 <- subset(vsd_habs.time, select=colData(vsd_habs.time)$time=="T4")
vsd_mat_habs.time.T4 <- assay(vsd_habs.time.T4)

# relabel data points for temperature
coldata_filtered.T4=subset(coldata_filtered, time=="T4")
coldata_filtered.T4$temperature=as.character(coldata_filtered.T4$temperature)
coldata_filtered.T4["temperature"][coldata_filtered.T4["temperature"] == "high"]<- "MHW"
coldata_filtered.T4["temperature"][coldata_filtered.T4["temperature"] == "low"]<- "Control"

p.T4 <- pca(vsd_mat_habs.time.T4, metadata = coldata_filtered.T4, removeVar = 0.1)
  screeplot(p.T4, axisLabSize = 18, titleLabSize = 22)
 tiff("PCA_T4.tiff", units="cm", width=15, height=12, res=300) 
  biplot(p.T4,
         x= "PC1", y="PC3",
    lab = paste0(p.T4$metadata$temperature),
    colby = 'level',
    colkey = c("none"="#41B3A3", 'low'= "#DDCC83", 'high'= "#E27D60" ),
    hline = 0, vline = 0,
    colLegendTitle = substitute(paste(italic('A. pacificum '), "concentration")),
    legendPosition = 'none',  axisLabSize = 12)
 dev.off()
  
  
  # Subset vsd matrix to T8 only
vsd_habs.time.T8 <- subset(vsd_habs.time, select=colData(vsd_habs.time)$time=="T8")
vsd_mat_habs.time.T8 <- assay(vsd_habs.time.T8)

# Relabel some datapoints
coldata_filtered.T8=subset(coldata_filtered, time=="T8")
coldata_filtered.T8$temperature=as.character(coldata_filtered.T8$temperature)
coldata_filtered.T8["temperature"][coldata_filtered.T8["temperature"] == "high"]<- "MHW"
coldata_filtered.T8["temperature"][coldata_filtered.T8["temperature"] == "low"]<- "Control"
p.T8 <- pca(vsd_mat_habs.time.T8, metadata = coldata_filtered.T8, removeVar = 0.1)
  screeplot(p.T8, axisLabSize = 18, titleLabSize = 22)
tiff("PCA_T8.tiff", units="cm", width=15, height=12, res=300) 
  biplot(p.T8,
         x= "PC2", y="PC3",
    lab = paste0(p.T8$metadata$temperature),
    colby = 'level',
    colkey = c("none"="#41B3A3", 'low'= "#DDCC83", 'high'= "#E27D60" ),
    hline = 0, vline = 0,
    legendPosition = 'none',
  axisLabSize = 12)
  dev.off()
  
```  
# Eigencorplot (using matrix from PCA plots)
```{r}   
# All timepoints
 tiff("eigencorplot_all timepoints.tiff", units="cm", width=30, height=20, res=300)
  eigencorplot(p,
               corMultipleTestCorrection = 'BH',
    metavars = c('time','level','temperature','replicate'), col = colorRampPalette(c("#41B3A3","white", "#E27D60" ))(100))
  dev.off()
  
  # T=4
   tiff("eigencorplot_T4.tiff", units="cm", width=30, height=20, res=300) 
  eigencorplot(p.T4,
    metavars = c('level','temperature','replicate'), col = colorRampPalette(c("#41B3A3","white", "#E27D60" ))(100))
  dev.off()
  
  
  # T=8
  tiff("eigencorplot_T8.tiff", units="cm", width=30, height=20, res=300) 
  eigencorplot(p.T8,
    metavars = c('level','temperature','replicate'), col = colorRampPalette(c("#41B3A3","white", "#E27D60" ))(100))
  dev.off()
  
```

#### Temp and time (different code)
```{r PCA, echo=FALSE, message=FALSE, warning=FALSE}
plot_data2=plotPCA(vsd_TT, intgroup=c("temperature", "time"), returnData=TRUE)
ggplot(plot_data2, aes(x=PC1, y=PC2, col=time, shape=temperature))+geom_point()

percentVar2 <- round(100 * attr(plot_data2, "percentVar"))
ggplot(plot_data2, aes(PC1, PC2, color=time, shape=temperature)) +
  geom_point(size=3) +
  ggtitle("Principal Component Analysis")+
  xlab(paste0("PC1: ",percentVar2[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar2[2],"% variance")) + 
  coord_fixed()+
  scale_color_manual(values = c("#41B3A3", "#DDCC83", 
                                "#E27D60", "#C38D9E"))+
  labs(color = "Time", shape="Temperature")
 
#Same thing different code 
 plotPCA(vsd_TT, intgroup=c("temperature", "time"))
 

# use prcomp to investigate other PCAs i.e. PC3 & PC4 for patterns
pca<-prcomp(t(vsd_mat))
df <- cbind(coldata_filtered, pca$x)
percentVar5<- round(100 * attr(pca, "percentVar"))
ggplot(df) + geom_point(aes(x=PC4, y=PC4, color = time, shape=replicate))+
  ggtitle("Principal Component Analysis")+
  xlab(paste0("PC3: ",percentVar5[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar5[2],"% variance")) + 
  coord_fixed()+
  scale_color_manual(values = c("#41B3A3", "#DDCC83","#99B7D4", 
                                "#E27D60", "#C38D9E", "#E79A20"))+
  labs(color = "time", shape="replicate")
```

###Heatmaps 
####Correlation heatmap

```{r Correlation heatmap, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
#compute correlation values, Use this to form a heatmap showing correlation between samples
# Note you can also use the rld to do this
vsd_cor_TT<- cor(vsd_mat)
head(vsd_cor_TT)

# Generate a condensed metadata file so that only a few things show up on heatmap
HM_df=coldata_filtered[-c(1,2, 4, 5, 7, 9, 10, 11, 12, 13)]
head(HM_df)
# This heatmap tells us that we have relatively high correlations >0.65
heatmap=pheatmap(vsd_cor_TT, annotation = HM_df)
heatmap

# Similar heatmap as above, without division into treatment, time and temp.
# Note that this takes into account the distances on the log transformed data, and t mean transpose 
# i.e. calc distance between data rows.
sampleDists = dist(t(assay(vsd_TT)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd_TT$level, vsd_TT$time, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix, intgroup=c("level", "time"),
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

## Same but using gplots
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(vsd_TT$level, vsd_TT$time, sep="-")
colnames(sampleDistMatrix) <- NULL

colours = colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
heatmap.2( sampleDistMatrix, trace="none", col=colours)
```

#### Heatmap of 20 most variable genes

```{r 20 most DE genes, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}}
# Select top 20 most DE genes
select <- order(rowMeans(counts(gene_counts_normalized)),
                decreasing=TRUE)[1:20]
topVarGenes <- head(order(-rowVars(assay(rld_TT))),50)
# Generate plot
df <- as.data.frame(colData(rld_TT)[,c("level", "time", "temperature")])
pheatmap(assay(rld_HTTint)[topVarGenes,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=TRUE, annotation_col= HM_df, main= "Heatmap of 20 most variable genes")
```

#### Heatmap using dds_TT, looking at all timepoints.

```{r Another heatmap, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# Select top 20 genes from vst matrix
topVarGenes <- head(order(rowVars(assay(vsd_TT)), decreasing = TRUE),35)
# Generate heatmap
my_colors=colorRampPalette( rev(brewer.pal(9, "Blues")) )(50)
mat <- assay(vsd_TT)[ topVarGenes, ]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(vsd_TT)[,c("level", "time")])
pheatmap(mat, annotation_col=df, scale = "row", cutree_cols = 4, cutree_rows = 3)

# Same but different heatmap function
heatmap.2( assay(vsd_TT)[ topVarGenes, ], scale="row",
trace="none", dendrogram="column",
col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255))

# Select top 20 genes from rld matrix
topVarGenes <- head(order(-rowVars(assay(rld_TT))),50)
# Generate heatmap
my_colors=colorRampPalette( rev(brewer.pal(9, "Blues")) )(50)
mat <- assay(rld_TT)[ topVarGenes, ]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(rld_TT)[,c("level", "time")])
pheatmap(mat, annotation_col=df, scale = "row", cutree_cols = 4, cutree_rows = 3)
```
#### Heatmap subsetting data, looking at only T=4.
```{r Another heatmap, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# Select for only T4
vsd_TT_T4=vsd_TT[, vsd_TT$time %in% c("T4")]
# Select top 20 genes from rld or vst matrix
topVarGenes_T4 <- head(order(-rowVars(assay(vsd_TT_T4))),50)
# Generate heatmap 
my_colors=colorRampPalette( rev(brewer.pal(9, "Blues")) )(50)
mat <- assay(vsd_TT_T4)[ topVarGenes_T4, ]
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(vsd_TT_T4)[,c("level", "temperature")])
pheatmap(mat, annotation_col=df, scale = "row", cutree_cols = 3, cutree_rows = 3)
```

## Differential gene expression
### Pairwise comparisons to generate lists of genes differentially expressed genes

```{r Exporting results to CSV files, message=FALSE, warning=FALSE, include=FALSE}
# Now we want to extract significantly differentially expressed genes from each of our results tables
# Will use a cutoff value of <0.05 padj value and a lfc of >0.58 (log2 of 1.5)
# Set thresholds
padj.cutoff <- 0.05
lfc.cutoff <- 0.58

# lfc.cutoff2<- 1.32
```

####Bulk comparisons 
#####T0 vs T4
```{r Timepoint comaprisons, message=FALSE, warning=FALSE, include=FALSE}
# Do bulk comparisons first 
# REMEMBER you can only do comparisons of the variables you put into the dds object design. 
# T0 vs T4
res_T0vT4= DESeq2:::results(dds_TT, contrast=c("time", "T4", "T0"), alpha = 0.05, lfcThreshold = 0.58)
head(res_T0vT4)
summary(res_T0vT4)
# Create a tibble of results
res_T0vT4_tb <- res_T0vT4 %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# Subset the tibble to keep only significant genes
sig_T0vT4 <- res_T0vT4_tb %>%
  dplyr::filter(padj < padj.cutoff & abs(res_T0vT4_tb$log2FoldChange) >lfc.cutoff )
sig_T0vT4
write.csv(as.data.frame(sig_T0vT4),file='sig_T0vT4.csv')
#108 DE genes

# SHRUNKEN DATA
resShrunk_T0vT4 <- lfcShrink(dds_TT, coef="time_T4_vs_T0", type="ashr")
# Create a tibble of results
resShrunk_T0vT4_tb <- resShrunk_T0vT4 %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# Subset the tibble to keep only significant genes with shrunken LFC to remove borderline DEs
sigShrunk_T0vT4 <- resShrunk_T0vT4_tb %>%
  dplyr::filter(padj < padj.cutoff & abs(resShrunk_T0vT4_tb$log2FoldChange) >lfc.cutoff )
sigShrunk_T0vT4
```
## Generate annotation file
```{r}
# Generate annotation file with a subset of the gene names (swissprot) annotations
annotation2 <- ProteinModels_swissprot_diamond %>% 
  dplyr::select(geneid, swissprot) %>% 
  dplyr::distinct()
# change column title to 'gene'
  colnames(annotation2)[1]="gene"

 # enerate annotation file with a subset of the gene names (diamond blast) annotations
annotation3 <- ProteinModels_diamond %>% 
  dplyr::select(geneid,diamond) %>% 
  dplyr::distinct()
# change column title to 'gene'
  colnames(annotation3)[1]="gene"  
  
#merge swissprot and diamond based on geneid, keep all data
# full_annot <- merge(annotation, annotation2, by.x = "gene", 
#              by.y = "gene", all.x = TRUE, all.y = TRUE)

full_annot <- merge(annotation2, annotation3, by.x = "gene", 
             by.y = "gene", all.x = TRUE, all.y = TRUE)


```

## Extracting lists of DEGs

```{r DEG extraction from tables, message=FALSE, warning=FALSE, include=FALSE}
 
# This merges the two tables based on their common identifier of "gene". It will delete any genes with no GOterm. 
sig_T0vT4=merge(sig_T0vT4, full_annot) 

## May not be necessary for you:
## This will collapse all the GO terms into one column, but keep different IDs distinct
# T0vT4_DE = sig_T0vT4 %>% 
#     dplyr::group_by(gene,log2FoldChange, padj) %>%
#     summarise(genename=paste(unique(genename), collapse = ", "))

```

### Heatmaps divided into subsets
This is not looking at absolute expression strength but at the most variable genes across the average of all samples.
Remember to use you vsd or rlog set to blind = FALSE
#### T=0 vs T=8 DE heatmap, both temps, all HAB concs
```{r DE genes heatmap, echo=FALSE, message=FALSE, warning=FALSE}
# Using results data from dds, data is not shrunken
# generate table of genes ordered by padj        
top30T0vT4=T0vT4_DE[order(T0vT4_DE$padj),]
#select the top 30
top30T0vT4=head(top30T0vT4, 30)
#convert to vector
top30T0vT4V=dplyr::pull(top30T0vT4, gene)

# generate matrix for heatmap
matT0vT4=assay(vsd_TT)
matT0vT4[top30T0vT4V,]

df <- as.data.frame(colData(vsd_TT)[,c("temperature", "habs", "time")])
pheatmap(matT0vT4[top30T0vT4V,], annotation_col=df,   scale = "row", cutree_cols = 2, 
         cutree_rows = 2, 
         fontsize_col = 7,          # column label font size 
         angle_col = 45, color=brewer.pal(11, "Spectral"))


# Remake the heatmap using shrunk data
resShrunk_T0vT4_DE=sig_ShrunkT0vT4 %>% 
 remove_rownames %>% column_to_rownames(var="gene")

# use same code as aboe to select top30 based on padj value      

df <- as.data.frame(colData(vsd_HTTint)[,c("temperature", "habs", "time")])
pheatmap(assay(vsd_HTTint)[DEs,], annotation_col=df, scale = "row",  cutree_cols = 3, 
         cutree_rows = 2, 
         fontsize_col = 7,          # column label font size 
         angle_col = 45, color=brewer.pal(11, "Spectral"))
```

###Venn diagrams 
####Method 1, using GOVenn

```{r GOVenn, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# This method of generating Venn diagrams requires a list of genes with their log2Fold changes so it can assign up and down regulation
# Here, you can combine lists of DE genes to look at cross over in comparisons. This is just an example with other data that hasn't been included to date. 
## Venn diagram using GOVenn
Venn_LTLH4= sig_LT_T4vsLTLH_T4 %>% dplyr::select(gene,log2FoldChange) 
Venn_LTHH4= sig_LT_T4vsLTHH_T4 %>% dplyr::select(gene,log2FoldChange) 
Venn_LT0vs4= sig_LT_T0vsLT_T4 %>% dplyr::select(gene,log2FoldChange) 
Venn_LT0vsLTLH4= sig_LT_T0vsLTLH_T4 %>% dplyr::select(gene,log2FoldChange) 
Venn_LT0vsLTHH4= sig_LT_T0vsLTHH_T4 %>% dplyr::select(gene,log2FoldChange) 
GOVenn(Venn_LT0vs4, Venn_LT0vsLTLH4)
GOVenn(Venn_LTLH4, Venn_LTHH4, label=c("LTLH", "LTHH"))

Venn_HTLH4= sig_HT_T4vsHTLH_T4 %>% dplyr::select(gene,log2FoldChange) 
Venn_HTHH4= sig_HT_T4vsHTHH_T4 %>% dplyr::select(gene,log2FoldChange)
GOVenn(Venn_HTLH4, Venn_HTHH4, label=c("HTLH", "HTHH"))
```

####Method 2, using ggVennDiagram

```{r ggVennDiagram, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# take the genes for each comparison and put them in a list format for ggvenndiagram 
# Take every gene table you wish to compare and convert to one list subsetted by sample ID and by gene
venn_id_T4=list(LTLH=(sig_LT_T4vsLTLH_T4$gene),
          LTHH=(sig_LT_T4vsLTHH_T4$gene),
          HTLH=(sig_HT_T4vsHTLH_T4$gene),
          HTHH=(sig_HT_T4vsHTHH_T4$gene))

venn_id_LT=list(HT=(sig_LTvHT$gene),
          LTLH=(sig_LTvLTLH$gene),
          LTHH=(sig_LTvLTHH$gene),
          HTLH=(sig_LTvHTLH$gene),
          HTHH=(sig_LTvLTHH$gene))


# Timepoint comparisons
venn_timepoint_comp=list(T0vT4=(sig_T0vT4$gene),
                    T4vT8=(sig_T4vT8$gene),
                    T0vT4=(sig_T0vT8$gene))

# HAB conc comparisons
venn_HAB_comp=list(noHABvlowHAB=(sig_noHABsvlowHABs$gene),
                    noHABvhighHAB=(sig_noHABsvhighHABs$gene))

# make a venn diagram using the lists
# Temp + HAB level comparisons
Venn_id_T4= ggVennDiagram(venn_id_T4,category.names = c("17°C + low HAB","17°C + high HAB", "22°C + low HAB", "22°C + high HAB"))+
  scale_x_continuous(expand = expansion(mult=.2))+
  scale_fill_distiller(palette="Spectral")+
  labs(title="Differentially expressed genes in all treatments at day 4")+
  theme(plot.title=element_text(hjust = 0.5))
Venn

# LT all timepoints
Venn_id_LT= ggVennDiagram(venn_id_LT,category.names = c("22°C", "17°C + low HAB","17°C + high HAB", "22°C + low HAB", "22°C + high HAB"))+
  scale_x_continuous(expand = expansion(mult=.2))+
  scale_fill_distiller(palette="Spectral")+
  labs(title="Differentially expressed genes in all treatments at day 4")+
  theme(plot.title=element_text(hjust = 0.5))
Venn_id_LT

# Timepoint comparisons
Venn_timepoint= ggVennDiagram(venn_timepoint_comp,category.names = c("T0 vs T4","T4 vs T8", "T0 vs T8"))+
  scale_x_continuous(expand = expansion(mult=.2))+
  scale_fill_distiller(palette="Spectral")+
  labs(title="Differentially expressed genes in all treatments at all three timepoints")+
  theme(plot.title=element_text(hjust = 0.5))
Venn_timepoint

# HAB conc comparisons
Venn_HAB= ggVennDiagram(venn_HAB_comp,category.names = c("no HAB v low HAB","no HAB vs high HAB"))+
  scale_x_continuous(expand = expansion(mult=.2))+
  scale_fill_distiller(palette="Spectral")+
  labs(title="Differentially expressed genes at different HAB concs at all three timepoints")+
  theme(plot.title=element_text(hjust = 0.5))
Venn_HAB
```

### Barplot of DEGs across each treatment
#### Plotting expression levels of specific genes of interest

```{r Barplot of significant genes, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# First identify genes of interest from your DEG tables
# Using the dds objects, can plot what's happening at gene level in each treatment
plotCounts(dds, gene="Scaffold_10__1_contigs__length_55049071.190", intgroup=c("treatment", "time"), returnData=TRUE) %>% 
  ggplot(aes("time", count)) + geom_boxplot(aes(fill=treatment))+ facet_wrap(~ time) + scale_y_log10()
```

```{r Barplot of significant genes, echo=FALSE, message=FALSE, warning=FALSE, fig.width = 10, fig.height=7}
# Use a dds object comparing HABs vs noHABs
# Using the dds objects, can plot what's happening at gene level in each treatment
topGene <- rownames(res_HABsvnoHABs)[which.min(res_HABsvnoHABs$padj)]
plotCounts(dds_HabsYN.time.temp, gene=topGene, intgroup=c("habs", "time"))
#output is scaffold_10__1_contigs__length_55049071.502
```

```{r}
# load in DEG values from csv
deg_comps2=read.csv("deg_comparisions2.csv", header = TRUE)
deg_comps2=deg_comparisions2
#For everything except multistressor
other=subset(deg_comps2, effect_of=="temperature_alone" |effect_of=="temperature_HAB" |effect_of== "hab_alone" |effect_of=="hab_temperature")
T4=subset(deg_comps2, timepoint=="4")
T8=subset(deg_comps2, timepoint=="8")
T4$id=factor(T4$id, levels=c("LTLH_4", "LTHH_4", "HTLH_4", "HTHH_4"))
T8$id=factor(T8$id, levels=c("LTLH_8", "LTHH_8", "HTLH_8", "HTHH_8"))

# change lebaels for facet plot
deg_comps2$timepoint <- factor(deg_comps2$timepoint, levels = c("0", "4", "8"),
                  labels = c("Day 0 (Initial conditions)", "Day 4 (post-HAB exposure)", "Day 8 (post-recovery)"))
deg_comps2$direction <- factor(deg_comps2$direction, levels = c("up", "down"),
                              labels = c("Upregulated", "Downregulated"))
deg_comps2$effect_of <- factor(deg_comps2$effect_of, levels = c("temperature_alone", "temperature_HAB", "hab_alone", "hab_temperature", "multistressor"),
                              labels = c("MHW (alone)", "MHW (A. pacificum present)", "A. pacificum (alone)","A. pacificum (MHW present)", "Multi-stressor"))

# make bar plot of DEGs from control T4 vs treatment T4 and the same for T8
tiff("DE gene expression in different treatments2.tiff", units="cm", width=40, height=30, res=300)
compbar=ggplot(deg_comps2, aes(x=treatment, y=degs, fill=direction))+
  theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'))+#transparent legend bg
  geom_bar(stat = "identity", position=position_dodge(), colour="black")+
    theme(legend.position = "bottom")+
  facet_wrap(~effect_of, scales = "free_x")+
   scale_x_discrete(breaks=c("high_temp.0", "high_temp.4", "high_temp.8", "high_temp_hab.4", "high_temp_hab.8", "low_temp_hab.4", "low_temp_hab.8", "high_temp_hab.4", "high_temp_hab.8", "high_temp_hab.4", "high_temp_hab.8"),labels=c("Day 0", "Day 4", "Day 8", "Day 4", "MHW (Day 8", "Day 4", "Day 8", "Day 4", "Day 8", "Day 4", "Day 8"))+
  theme(axis.text.x = element_text(size=11, angle = 45, vjust = 1, hjust=1), axis.title.y = element_text(size=12), 
        axis.title.x =element_text(size=12, vjust=0), axis.text.y = element_text(size=11), plot.title = element_text(size=14))+
  scale_fill_manual("Legend", values = c("Upregulated" = "#E27D60", "Downregulated" = "#41B3A3"))+
  labs(x="Treatment", y="Number of differentially expressed genes")+
  ggtitle("Differential gene expression based on applied stressor")
    compbar 
  dev.off()
    # ggsave('DEG_graph_transparent.png', compbar, bg='transparent')
   
  ##Subset data to be for multistress only
multi=subset(deg_comps2, effect_of=="multistressor")   
   
multi$timepoint <- factor(multi$timepoint, levels = c("4", "8"),
                  labels = c("Day 4 (post-HAB exposure)", "Day 8 (post-recovery)"))
multi$direction <- factor(multi$direction, levels = c("up", "down"),
                              labels = c("Upregulated", "Downregulated"))
multi$effect_of <- factor(multi$effect_of, levels = c("multistressor"),
                              labels = c("Multi-stressor"))

   ## Multistressor only 
tiff("DE gene expression in multistressor treatments.tiff", units="cm", width=15, height=25, res=300)
   compbar2=ggplot(multi, aes(x=treatment, y=degs, fill=direction))+
  theme(
    panel.background = element_rect(fill='transparent'), #transparent panel bg
    plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
    panel.grid.major = element_blank(), #remove major gridlines
    panel.grid.minor = element_blank(), #remove minor gridlines
    legend.background = element_rect(fill='transparent'))+#transparent legend bg
  geom_bar(stat = "identity", position=position_dodge(), colour="black")+
    theme(legend.position = "none")+
  theme(axis.text.x = element_text(size=18, angle = 45, vjust = 1, hjust=1), axis.title.y = element_text(size=18), 
        axis.title.x =element_text(size=16, vjust=0), axis.text.y = element_text(size=16), plot.title = element_text(size=14))+
      scale_x_discrete(breaks=c("high_temp_hab.4", "high_temp_hab.8"),labels=c("Multistressor day 4", "Multistressor day 8"))+ 
  scale_fill_manual("Legend", values = c("Upregulated" = "#E27D60", "Downregulated" = "#41B3A3"))+
     labs(y="Number of differentially expressed genes")
    compbar2
    dev.off()

```
## Ridge plot
```{r}
##For more info:  https://bioinformatics.ccr.cancer.gov/docs/btep-coding-club/CC2023/FunctionalEnrich_clusterProfiler/

# SET THE DESIRED ORGANISM HERE
organism = "org.Hs.eg.db"
library(organism, character.only = TRUE)
keytypes


# reading in data from deseq2
df = read.csv("T4 HAB v noHAB both temps.csv", header=TRUE)

# we want the log2 fold change 
original_gene_list <- df$log2FoldChange
original_gene_list
# name the vector
names(original_gene_list) <- df$swissprot
names
# omit any NA values 
gene_list<-na.omit(original_gene_list)
gene_list
# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)
gene_list
keytypes(org.Hs.eg.db)
gse_BP <- gseGO(geneList=gene_list, 
             ont ="BP", 
             keyType = "UNIPROT", 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")
gse_BP

gse_MF <- gseGO(geneList=gene_list, 
                ont ="MF", 
                keyType = "UNIPROT", 
                minGSSize = 3, 
                maxGSSize = 800, 
                pvalueCutoff = 0.05, 
                verbose = TRUE, 
                OrgDb = organism, 
                pAdjustMethod = "none")

gse_CC <- gseGO(geneList=gene_list, 
                ont ="CC", 
                keyType = "UNIPROT", 
                minGSSize = 3, 
                maxGSSize = 800, 
                pvalueCutoff = 0.05, 
                verbose = TRUE, 
                OrgDb = organism, 
                pAdjustMethod = "none")

gse_all <- gseGO(geneList=gene_list, 
                ont ="all", 
                keyType = "UNIPROT", 
                minGSSize = 3, 
                maxGSSize = 800, 
                pvalueCutoff = 0.05, 
                verbose = TRUE, 
                OrgDb = organism, 
                pAdjustMethod = "none")

# Make the ridge plots
MF1= ridgeplot(gse_MF, showCategory = 12) + 
  labs(x = "Enrichment distribution")
tiff("HABvnoHAB T4 RP_MF.tiff", units="cm", width=15, height=20, res=300)
MF1 + theme(axis.text.y=element_text(size=10))
dev.off()

BP1=ridgeplot(gse_BP, showCategory = 12) + labs(x = "Enrichment distribution") 
tiff("HABvnoHAB T4 RP_BP.tiff", units="cm", width=15, height=20, res=300)
BP1 + theme(axis.text.y=element_text(size=10))
dev.off()


CC1= ridgeplot(gse_CC, showCategory = 12) + labs(x = "Enrichment distribution")
dev.off()
tiff("HABvnoHAB T4 RP_CC.tiff", units="cm", width=15, height=20, res=300)
CC1 + theme(axis.text.y=element_text(size=10))
dev.off()


All_1= ridgeplot(gse_all, showCategory = 20, label_format = 60) + labs(x = "Enrichment distribution")
tiff("HABvnoHAB T4 RP_all.tiff", units="cm", width=30, height=28, res=300)
All_1 + theme(axis.text.y=element_text(size=16))
dev.off()


```

```{r, session info}
sessionInfo()
```

Note that the `echo = FALSE` parameter was added to the code chunk to
prevent printing of the R code that generated the plot.
